app/subscriber.py

```diff
@@ -127,6 +120,99 @@ def offline_receipt_to_case(message: Message):
     log.info('Message processing complete')
 
 
+def ppo_undelivered_mail_to_case(message: Message):
+    log = logger.bind(message_id=message.message_id,
+                      subscription_name=PPO_UNDELIVERED_SUBSCRIPTION_NAME,
+                      subscription_project=PPO_UNDELIVERED_SUBSCRIPTION_PROJECT_ID)
+
+    log.debug('Pub/Sub Message received for processing')
+
+    payload = validate_message(message.data, log, ['caseRef', 'productCode'])
+    if not payload:
+        return  # Failed validation
+
+    case_ref, product_code, date_time = payload['caseRef'], payload['productCode'], payload['dateTime']
+
+    log = log.bind(case_ref=case_ref, created=date_time, product_code=product_code)
+
+    receipt_message = {
+        'event': {
+            'type': 'UNDELIVERED_MAIL_REPORTED',
+            'source': 'RECEIPT_SERVICE',
+            'channel': 'PPO',
+            'dateTime': date_time,
+            'transactionId': str(uuid.uuid4())
+        },
+        'payload': {
+            'fulfilmentInformation': {
+                'caseRef': case_ref,
+                'productCode': product_code
+            }
+        }
+    }
+
+    send_message_to_rabbitmq(json.dumps(receipt_message), routing_key=UNDELIVERED_MAIL_ROUTING_KEY)
+    message.ack()
+
+    log.debug('Message processing complete')
+
+
+def qm_undelivered_mail_to_case(message: Message):
+    log = logger.bind(message_id=message.message_id,
+                      subscription_name=QM_UNDELIVERED_SUBSCRIPTION_NAME,
+                      subscription_project=QM_UNDELIVERED_SUBSCRIPTION_PROJECT_ID)
+
+    log.debug('Pub/Sub Message received for processing')
+
+    payload = validate_message(message.data, log, ['questionnaireId'])
+    if not payload:
+        return  # Failed validation
+
+    questionnaire_id, date_time = payload['questionnaireId'], payload['dateTime']
+
+    log = log.bind(questionnaire_id=questionnaire_id, created=date_time)
+
+    receipt_message = {
+        'event': {
+            'type': 'UNDELIVERED_MAIL_REPORTED',
+            'source': 'RECEIPT_SERVICE',
+            'channel': 'QM',
+            'dateTime': date_time,
+            'transactionId': str(uuid.uuid4())
```

Comment:
we shouldn't be creating a new transactionId here, there should be one coming in on the message from GCP pub/sub message

Reply:
not present according to the spec. https://collaborate2.ons.gov.uk/confluence/pages/viewpage.action?pageId=20189919